Extend NemoLite2D manual versions

sergisiso opened the pull request:
This PR tries to extend and improve the manual implementations from a top level point of view, this is achieved by:
- [x] Create a top-level Makefile and namelist so that all tests can be compiled and executed with exact parameters easily.
- [x] All implementations are in a separate directory using the psykal_x and single_file_x name convention. ~~Also add placeholders for OmpSs, SYCL, ...~~ see [README](https://github.com/stfc/PSycloneBench/blob/extend_manual_versions/ocean/nemo/nemolite2d/manual_versions/README.md)
- [x] Some compiler parameters are hardcoded on the Makefiles, it would be good to move this to the top-level `compiler_setup` files.
- [x] Each manual_version Makefile replicates a lot of logic that some of them get out of date. Make a top-level `common.mk` with the common Makefile logic and include the file from each Makefile
- [x] At the end I would like the Makefile recuses to all implementations an easily produce comparable results, At the moment I get:
```
manual_versions $ make run
Implementation           ua checksum     uv checksum     time-stepping
psykal_serial:          0.13226697E+02  0.14386789E+03  0.13026E+01
psykal_omp:             0.13226697E+02  0.14386789E+03  0.57432E+00
```

sergisiso commented:
@arporter In the psykal_omp manual implementation there is a paraview.pvsm file that I initially deleted thinking it was probably the remains of some old output. But looking at it again (at its textual representation, I don't have paraview installed at the moment) I am wondering if this was a 'state file' to visualize the output of other files instead? If thats the case, is there any reason this file is only inside the openmp implementation and not the others? And does it was some restrictions? It seems some values/sizes are hardcoded in the file.

arporter commented:
I think that this file was a leftover from the very first implementation by Hedong at NOC. I think it was a 'state file' but I've always just used gnuplot to visualise the results. We can remove it.

sergisiso commented:
@arporter This PR is finally ready for review. As mentioned before the main objective was to simplify and unify the build system of the different versions so it is easy to maintain and to check the performance in a equivalent way (same compiler flags, same parameters, ...)

I also bring the dl_esm_inf and FortCL to master, which allows to run MPI correct checksums and JIT OpenCL.

A good way to test that all versions compile and run is to execute `make` and `make summary_table` from `benchmarks/nemo/nemolite2d/manual_versions`. I tested it with 2 compiler toolchains in Wheatley and I get:
```
Implementation          Compiler        Arch    ua checksum     uv checksum     time/step
psykal_serial:          gcc             cpu     0.52670513E+02  0.57100321E+03  0.20225E+00
psykal_omp:             gcc             cpu     0.52670513E+02  0.57100321E+03  0.66258E-01
psykal_dm:              gcc             cpu     0.52670513E+02  0.57100321E+03  0.456806E-01
psykal_serial:          icc             cpu     0.52670513E+02  0.57100321E+03  0.90677E-01
psykal_omp:             icc             cpu     0.52670513E+02  0.57100321E+03  0.47225E-01
psykal_dm:              icc             cpu     0.52670513E+02  0.57100321E+03  0.654124E-01
```
Interestingly intel gets better serial and openmp performance but not so much MPI performance, by contrast gnu compiler is best with MPI.

I also refactored the OpenCL implementations, basically the `opencl/fortran` is now `psykal_opencl` (and it uses FortCL similar to what PSyclone generates) instead of its own copies of clfortran. The old `opencl/c` is now `single_file_c_opencl`. The OpenCL built system are not as polished as the rest (I still get complains from the encoding from time to time and not all Makefiles are refactored to use the OpenCL compier_vars). But since I would like to merge this Makefile structure already and I am the main user of OpenCL at the moment, I propose to finish polishing the OpenCL Makefiles in a future PR .

arporter commented on compiler_setup/intel.sh:
```diff
@@ -2,7 +2,7 @@
 F90=ifort
```
I guess the various `compiler_setup/*.sh` files should have copyright/licence notices too.

sergisiso commented on README.md:
```diff
@@ -38,20 +38,9 @@ by Hedong Liu of the National Oceanography Centre, Liverpool, UK.
-Please see the README.md files in each of the directories for more
-information.
+and Kernel layers but the PSy layer has been written manually.
+See the [README.md](ocean/nemo/nemolite2d/manual_versions/README.md)
```
Fixed

sergisiso commented on README.md:
```diff
@@ -38,20 +38,9 @@ by Hedong Liu of the National Oceanography Centre, Liverpool, UK.
-information.
+and Kernel layers but the PSy layer has been written manually.
+See the [README.md](ocean/nemo/nemolite2d/manual_versions/README.md)
+file in the NemoLite2D manual`_versions directory for more information.
```
Fixed

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/Makefile:
```diff
@@ -0,0 +1,35 @@
+# This Makefile iterates through the IMPLEMENTATION directories and
+# compiles, executes and prints a summary with a copy of an equivalent
```
Replaced with suggested sentence

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/README.md:
```diff
@@ -1,29 +1,47 @@
 # Manual versions of NEMOLite2D #
 
-## single_file_acc ##
+The manual implemented versions of NemoLite2D are divided in 2 subgroups:
```
Done (but without hyphen)

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/README.md:
```diff
@@ -1,29 +1,47 @@
 
-OpenACC versions of the single source-file form of NEMOLite2D with
-different optimisations applied.
+1. The "PSyKAl" implementations, which have each layer the applications
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/README.md:
```diff
@@ -1,29 +1,47 @@
+already separated using the PSyKAl model but the ParallelSystem is
+implemented manually. In these implementations usually the Algorithm
+can be found in the `nemolite2d.f90` file, the manual PSy-layer is found
+on the `time_step.f90` and the kernels which are encoded as independent
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/README.md:
```diff
@@ -1,29 +1,47 @@
+implemented manually. In these implementations usually the Algorithm
+can be found in the `nemolite2d.f90` file, the manual PSy-layer is found
+on the `time_step.f90` and the kernels which are encoded as independent
+functions on the `time_step.f90` or the `../kernels` are used. These are:
```
Oops this was not very well written, fixed now.

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/README.md:
```diff
@@ -1,29 +1,47 @@
-## psykal_omp ##
 
-OpenMP version of the PSyKAl form of NEMOLite2D.
+2. The "single\_file" versions that due to programming model limitations
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/README.md:
```diff
@@ -1,29 +1,47 @@
 
-OpenMP version of the PSyKAl form of NEMOLite2D.
+2. The "single\_file" versions that due to programming model limitations
+or code simplification the implementation is not separated into the PSyKal
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/README.md:
```diff
@@ -1,29 +1,47 @@
+## Makefile ##
+A top-level Makefile is provided to recurse down to the multiple
+implementations, compile them and execute them using a common set of
+parameters (provided by the lop-level namelist).
```
Fixed

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/common.mk:
```diff
@@ -0,0 +1,64 @@
+INF_DIR = ${SHARED_DIR}/dl_esm_inf/finite_difference
+INF_INC = ${INF_DIR}/src
+INF_LIB = ${INF_DIR}/src/lib_fd.a
+# common
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/common.mk:
```diff
@@ -0,0 +1,64 @@
+	ln -sf ../../kernels/fortran/$@ .
+
+# Common rules
+timer_lib:
```
They were is some sub-folders, I added it again in common.mk

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_dm/Makefile:
```diff
@@ -1,90 +1,40 @@
-# export F90=mpif90
-# export F90FLAGS="-O3"
-# export AR=ar
+# environment variables.
```
Fixed

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_dm/Makefile:
```diff
@@ -1,90 +1,40 @@
-# API lib is an archive that must come at the end of the list of objects
-# passed to the linker
-COMMON_MODULES = $(MODULES) ${COMMON_LIB} ${INF_LIB}
+# The modules that are common to both targets (API lib is an archive that
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_dm/namelist:
```diff
@@ -1,15 +1,15 @@
 !-----------------------------------------------------------------------
```
The idea was that one could work from the top-level directory if intends to benchmark across versions (then namelist is copied), or from inside a directory if he/she is just working/altering a particular version (then namelist is individual to each implementation).

I suppose we could always tell the Makefile to copy the namelist from the top-level if it is not found inside the folder if that is something you would prefer?

sergisiso replied:
Changed subfolder namelists to symbolic links

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_omp/namelist:
```diff
@@ -1,15 +1,15 @@
 !-----------------------------------------------------------------------
-   jpiglo      =     258               !  number of columns of model grid
-   jpjglo      =     258               !  number of rows of model grid
+   jpiglo      =    1024               !  number of columns of model grid
```
See above

sergisiso replied:
Changed subfolder namelists to symbolic links

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/Makefile:
```diff
@@ -0,0 +1,36 @@
+LDFLAGS += ${OMPFLAGS}
+FFLAGS += ${OPENCL_INCLUDE}
+
+.PHONY: all nemolite2d.exe device_binary
```
Removed

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/Makefile:
```diff
@@ -0,0 +1,36 @@
+	${F90} -o $@ ${COMMON_MODULES} ${OPENCL_LIBS} ${COMMON_LIB} \
+        ${FCL_LIB} ${INF_LIB} ${TIMER_LIB} -lOpenCL $(LDFLAGS)
+
+device_binary: %.cl
```
I added a comment explaining its purpose and how to use it. However, this target is the one I need to play a bit more (after this PR) in order to be confident it works across different target accelerators/compilers. So maybe it will be polished in the near future.

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/README.md:
```diff
@@ -0,0 +1,43 @@
+settings for the Gnu compiler suite (with JIT OpenCL).
+If you are using Bash and the Gnu compiler then:
+
+    > . ../../../../../compiler_setup/gnu.sh
```
It is here, but it wasn't in the sentence above :)

Also I replaced `.` with `source` for better readability.

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/README.md:
```diff
@@ -0,0 +1,43 @@
+[nitend], whether or not and how often to do output [irecord]) may be
+configured by editing the `namelist` file.
+
+Additionally the OpenCL execution needs that the environment variable
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/README.md:
```diff
@@ -0,0 +1,43 @@
+PSYCLONE_KERNELS_FILE is set up to reference the target device binary
+or the target source code. If the source code is given, the OpenCL
+runtime system will JIT compile the necessary device objects at runtime.
+The following command shows and example of a JIT compiled execution:
```
Oops fixed

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/nemolite2d.f90:
```diff
@@ -0,0 +1,152 @@
+  ! points immediately to the North and East of a T point all have the
+  ! same i,j index).  This is the same offset scheme as used by NEMO.
+  model_grid = grid_type(GO_ARAKAWA_C, &
+  !  BC_PERIODIC, BC_NON_PERIODIC ??
```
Deleted

sergisiso replied:
Also deleted the some line in the other implementations.

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/nemolite2d.f90:
```diff
@@ -0,0 +1,152 @@
+  TYPE(r2d_field), INTENT(INOUT) :: hu, hv, ht
+  CALL invoke_time_step(ssha_t, sshn_t, sshn_u, sshn_v, hu, hv, un, vn, ua, ht, ssha_u, va, ssha_v, istp)
+
+!  call model_write(grid, istp, ht, sshn, un, vn)
```
done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/nemolite2d.f90:
```diff
@@ -0,0 +1,152 @@
+  TYPE(r2d_field), INTENT(INOUT) :: un, vn, sshn_t, sshn_u, sshn_v
+  TYPE(r2d_field), INTENT(INOUT) :: ua, va, ssha_t, ssha_u, ssha_v
+  TYPE(r2d_field), INTENT(INOUT) :: hu, hv, ht
+  CALL invoke_time_step(ssha_t, sshn_t, sshn_u, sshn_v, hu, hv, un, vn, ua, ht, ssha_u, va, ssha_v, istp)
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/time_step_mod.f90:
```diff
@@ -0,0 +1,922 @@
+
+    SUBROUTINE invoke_time_step(ssha_t, sshn_t, sshn_u, sshn_v, hu, hv, un, &
+                                vn, ua, ht, ssha_u, va, ssha_v, istp)
+        USE field_mod
```
Done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/time_step_mod.f90:
```diff
@@ -0,0 +1,922 @@
+            CALL add_kernels(11, kernel_names)
+        END IF
+    END SUBROUTINE psy_init
+
```
yep :)

sergisiso commented on Makefile:
```diff
@@ -39,21 +39,19 @@ nemolite_acc:
 
 clean:
```
The `$@` for passing the clean/allclean target seems to work.

sergisiso replied:
Aha, I didn't notice the multiple target labels, this is a good idea :)

(I haven't done the foreeach, I think the list of instructions looks nicer here)

sergisiso commented on Makefile:
```diff
@@ -39,21 +39,19 @@ nemolite_acc:
 	${MAKE} -C ./benchmarks/shallow/SEQ/original allclean
 	${MAKE} -C ./benchmarks/shallow/OMP allclean
-
+	${MAKE} -C ./benchmarks/smallmatvec/manual_versions/openmp clean
```
Well, the smallmatvec Makefile didn't have an `allclean` target, but now I added it (even if just does the same as clean for now)

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/allkernels.cl:
```diff
@@ -0,0 +1,532 @@
+  va[jj * vaLEN1 + ji] = ((((vn[jj * vnLEN1 + ji] * (hv[jj * hvLEN1 + ji] + sshn_v[jj * sshn_vLEN1 + ji])) + ((rdt * (((adv + vis) + cor) + hpg)) / e12v[jj * e12vLEN1 + ji])) / (hv[jj * hvLEN1 + ji] + ssha_v[jj * ssha_vLEN1 + ji])) / (1.0 + (cbfr * rdt)));
+}
+
+__kernel void next_sshu_code(
```
:)

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_dm/Makefile:
```diff
@@ -1,90 +1,40 @@
-SHARED_DIR = ../../../../../shared
+# Make sure that F90 has a MPI compiler selected.
+ifeq (,$(findstring mpi,$(F90)))
+	F90 := mpif90 -fc=${F90}
```
Ok, I tried to make it more robust (albeit also uglier as I can't indent the Makefile conditionals).
Now it aborts if there is no `mpifort` but if it exist, it still tries to use the compiler given by the loaded compiler_vars, unless a custom mpi F90 is given. Does this make sense?

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/Makefile:
```diff
@@ -0,0 +1,36 @@
+# Makefile for manual OpenCL version of PSyKAl'd NEMOLite2D.
+#
+# This Makefile picks up the compiler to use any flags from
```
I briefly mention what are the expected environment variables, but I will probably extend/modify this when I have the time to try it with more OpenCL targets/compilers.

sergisiso commented:
@arporter Ready for the next review.

It seems last time the MPI implementation didn't work for you as I didn't consider OpenMPI and MPICH families have different ways to specify which base compiler to use. Hopefully this is fixed now.

Also, I addressed all comments but the license ones. As you mention there is already a top-level license, and it seems other Makefiles and sh-scripts that I haven't touched don't have the copyright notice, so it may be better to keep them the same.

arporter commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/nemolite2d.f90:
```diff
@@ -44,7 +44,6 @@ program gocean2d
   ! same i,j index).  This is the same offset scheme as used by NEMO.
   model_grid = grid_type(GO_ARAKAWA_C, &
-  !  BC_PERIODIC, BC_NON_PERIODIC ??
                          (/GO_BC_EXTERNAL,GO_BC_EXTERNAL,GO_BC_NONE/), &
```
Please add an attribution to Hedong Liu, NOC in the model docstring above (line 12).

sergisiso replied:
Done

arporter commented:
Almost there now. For some reason GitHub wasn't letting me continue with existing conversations but re. the namelist files, I think it would be best if we just have one and copy it in/sym link it in the lower folders.

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/Makefile:
```diff
@@ -11,6 +11,7 @@ IMPLEMENTATIONS += psykal_dm
 # psykal_ompss and psykal_acc have not been adapted to this Makefile format
 # yet.
 
+.PHONY: clean allclean compile summary_table
```
done

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_dm/Makefile:
```diff
@@ -1,11 +1,27 @@
+# If F90 already has a custom mpi compiler command, it will use it,
+# otherwise it will form a mpich or openmpi -family command using the
+# compiler defined in F90.
+ifndef F90
```
I added comments to the relevant conditions, hopefully it is more readable now.

arporter replied:
Thanks.

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/README.md:
```diff
@@ -1,13 +1,15 @@
+with the Parallel System layer implemented manually.
+In these implementations usually the Algorithm can be found in the
+`nemolite2d.f90` file, the manual PSy-layer is found in `time_step.f90` and
+the kernels which are encoded as independent functions in `time_step.f90` are
```
I removed this part to have a simpler sentence and just going to the point of where things are located.

sergisiso commented on benchmarks/nemo/nemolite2d/manual_versions/psykal_opencl/nemolite2d.f90:
```diff
@@ -0,0 +1,152 @@
+program gocean2d
```
Done

sergisiso commented:
@arporter PR ready for next review

sergisiso commented:
I remove the Ready for review label because:
- [ ] The last changes cause an error in `psykal_serial` standalone compilation (due to wrong order)
- [ ] I want to include the last dl_esm_inf PR which would allow for OpenCL to be included in summary_tables as it will report checksums

sergisiso commented:
@arporter Ready for review again.

arporter commented:
All requested changes have been made. I get summary performance results for serial, OMP and DM on my laptop without problems :-) Would be nice to try the OpenCL but I'll wait for your next PR for that :-)
